echidna/core/fill_spectrum.py

```diff
@@ -3,432 +3,299 @@
+                fill_kwargs[extractor._name] = extractor.get_value_root(mc)
+            else:
+                fill = False
+                break
+        if fill:
             try:
-                spectrum.fill(energy, position, time, 1.)
+                spectrum.fill(**fill_kwargs)
+                spectrum._raw_events += 1
             except ValueError:
                 pass
 
-    return spectrum
-
 
-def fill_mc_ntuple_spectrum(filename, T, spectrumname="", spectrum=None):
-    """**Weights have been disabled.**
-    This function fills in the ndarray of energies, radii, times and
-    weights. It takes the true quenched energies and positions of the
-    events from ntuple. In order to keep the statistics, the time
-    dependence is performed via adding weights to every event depending
-    on the time period. Both, studied time and Half-life must be
-    written in the same units.
+def fill_from_root(filename, spectrum_name="", config=None, spectrum=None,
+                   bipo=False, **kwargs):
+    """ This function fills in the ndarray (dimensions specified in the config)
+    with weights. It takes the parameter specified in the config from the
+    events in the root file.
 
     Args:
-      filename (str): The ntuple to study
-      T (float): The Half-life of a studied background
-      spectrumname (str, optional): A name of future spectrum. Not
+      filename (str): A root file to study
+      spectrum_name (str, optional): A name of future spectrum. Not
         required when appending a spectrum.
+      config (:class:`echidna.core.spectra.SpectraConfig` or string, optional):
+        The config or directory to the config for the spectrum.
+        Not requried when appending a spectrum.
       spectrum (:class:`echidna.core.spectra.Spectra`, optional):
         Spectrum you wish to append. Not required when creating a
         new spectrum.
+      bipo (bool, optional): Applies the bipo cut if set to True.
+        Default is False.
+      kwargs (dict): Passed to and checked by the dsextractor.
 
     Raises:
-      ValueError: If spectrumname is not set when creating a new
+      ValueError: If spectrum_name is not set when creating a new
         spectrum.
+      IndexError: Unknown dimension type (not mc, truth or reco).
 
     Returns:
-      spectrum (:class:`echidna.core.spectra.Spectra`)
+      :class:`echidna.core.spectra.Spectra`: The filled spectrum.
     """
-    print filename
-    chain = TChain("output")
-    chain.Add(filename)
+    if type(config) == str:
+        config = spectra.SpectraConfig.load_from_file(config)
     if spectrum is None:
-        if spectrumname == "":
+        if spectrum_name == "" or not config:
             raise ValueError("Name not set when creating new spectra.")
-        spectrum = spectra.Spectra(str(spectrumname), 10.*chain.GetEntries())
+        spectrum = spectra.Spectra(str(spectrum_name),
+                                   dsreader.GetEntryCount(),
```

Comment:
One final thing on this module, dsreader is undefined here, you define it below on line 277.